### ISSUE TITLE
Distinguishing "development" from "production"

### ISSUE THREAD

**Issue by zeke** [electron/electron#7714]
## The Problem

When developing an Electron app, it is sometimes necessary to change the app's behavior depending on whether the app is being run in development or is packaged and running on an end-user's machine. Examples include:
- Enabling DevTools in "development", disabling in "production"
- Changing logging/debugging behavior
- Using different endpoints for talking to webservices
- Enabling/disabling caching
## Proposal

Electron should provide an intuitive way for users to detect (from any process) whether or not their app is running in a packaged state.
#### `process.defaultApp`

Electron has a property called `process.defaultApp` which is set to `true` when running Electron with a target directory, e.g. `electron .`. There are at least two problems with this:
1. Only works in the main process. Rendered processes can't access this property by default.
2. The name `defaultApp` is not intuitive, and the [docs give no indication](http://electron.atom.io/docs/all/#processdefaultapp) of this property's practical use.
#### `electron-is-dev`

[electron-is-dev](https://github.com/sindresorhus/electron-is-dev/) is a userland module that solves this problem. This modules is great, but the issue here is discoverability: Users have to know about it before they can use it.
#### `ELECTRON_ENV`

See https://github.com/electron-userland/electron-prebuilt/pull/118, an open PR that aims to set `process.env.ELECTRON_ENV` to `development` in development.

One might argue that this technique could be used by simply prefixing the environment variable to a start script:

``` json
{
  "scripts": {
    "start": "ELECTRON_ENV=development electron ."
  }
}
```

Unfortunately this is not a cross-platform solution: it won't work on Windows without bringing in an external dependency like [cross-env](https://github.com/kentcdodds/cross-env).
## What to do?

I'm in favor of the `process.env.ELECTRON_ENV` approach, setting it to `development` when run with `electron some/app/dir`, and leaving it `undefined` for packaged apps.

I like this solution because it resembles the `NODE_ENV` convention from Node.js, and gives users a familiar way to override the value easily, if needed. I would also like to solve this problem in Electron itself, rather than `electron-prebuilt`, so all developers will experience the same behavior, regardless of how they're running Electron.

Before jumping into implementation, I would love feedback from users. Does this approach sound good?

cc @parro-it @sindresorhus @ungoldman @kevinsawicki

**Comment by baconface** [comment]
For us it has always been easy to do with minimal code. This solution works in both the main and render processes.

Here is an example:
```javascript
function isDev() {
  return process.mainModule.filename.indexOf('app.asar') === -1;
}
```

Usage:
```javascript
if (isDev) {
  // Do dev stuff
}
```

**Comment by zeke** [comment]
@baconface that's a nice one-liner. I see a few drawbacks, though:

1. Requires deep knowledge of how Electron works. Not intuitive to new users.
1. Will break for apps that are not using ASAR.

**Comment by baconface** [comment]
@zeke I agree. And even though this works for us I would also like to see an official solution as well.

**Comment by ungoldman** [comment]
Hi @zeke! I agree with your point about solving the issue in electron rather than userland, and think the `NODE_ENV` convention is a good model. I also think it's important to have something distinct from `NODE_ENV`, so `ELECTRON_ENV` sounds good to me. So basically yes, I would like to see `ELECTRON_ENV` set to `development` when an app is unpackaged. 👍

**Comment by zeke** [comment]
@ungoldman glad to hear it.

> I also think it's important to have something distinct from `NODE_ENV`

I forgot to mention that during an internal discussion, `NODE_ENV` was floated as an alternative to `ELECTRON_ENV`. I tend to agree with you, but I wonder: any particular reason you think the two should be distinct?

**Comment by ungoldman** [comment]
> any particular reason you think the two should be distinct?

Because it's not node? I think it's possible to have a situation where NODE_ENV is already set in the environment... but this may just be me imagining things.

**Comment by zeke** [comment]
> Because it's not node?

👍

**Comment by baconface** [comment]
I don't think a distinction is appropriate. Code is not ran in the Electron context. It is ran in in the Node context (although a slightly modified Node). Accessing Electron's functionalities are APIs available in that Node context. Using NODE_ENV would also yield useful for modules that use it already for debugging/development.

**Comment by MarshallOfSound** [comment]
I put some thought into this a while back as well, internally in GPMDP we use a rather funky `--dev` CLI arg to switch between "production" and "development" modes.  I would much prefer a standard way in Electron to do it but I'd like to make a case for **not** using an environment variable.

1. Using environment variables in standard node modules is OK, in reality most node modules are developed and run on *nix based systems that allow commands such as `NODE_ENV=production npm run doThing`.  However Electron has to remember that it is completely cross-platform and as such needs an easy way for Windows users to make this switch as well.  Unfortunately windows does not support that environment variable syntax and requires that you use another module `cross-env`
2. All core features of Electron are currently toggled through an API and I believe that should remain the same, we have some debugging environment variables but nothing that modifies how code actually runs.  I believe if users want an environment variable to switch between production and development they can configure that themselves.

I would love to see an api such as

```markdown
## Events
### Event: 'in-production-changed'

Returns:

* `event` Event
* `inProd` Boolean

Emitted when the application is switched in or out of production mode


## Methods
### `app.inProduction()`

Returns `Boolean` - Whether the application should run in a production environment

### `app.setProduction(inProd)`

* `inProd` Boolean - Whether the application should run in a production environment

**NOTE:** This does not affect any of Electron's internals, only your code and third
party modules that read this option.
```

This way if a user does want to map `ELECTRON_ENV` to `inProduction()` they simple have to call

```js
app.setProduction(!!process.env.ELECTRON_ENV);
```

With this API renderer modules can use `remote.app.inProduction()` to obtain the same information.  What do people think of this (relatively simple to implement) approach.

**Comment by enlight** [comment]
I could see some utility in providing a standard way of determining if an app is packaged or not, but this issue seems to be conflating that feature with app environments/configurations. I don't think Electron should assume anything about how apps handle different environments, apps may have more than one development environment and/or more than one production environment.

**Comment by zeke** [comment]
> I could see some utility in providing a standard way of determining if an app is packaged

Agreed. There is less ambiguity in answering whether an app is "packaged". This approach would also eliminate the need for users to set the value. Hence we could just get away with a single getter method. This is also more clear than the existing main-process `process.defaultApp` property, which is not intuitively named.

Several existing boolean-returning API methods have an `is` prefix:

app.isReady()
app.isUnityRunning()
app.isAccessibilitySupportEnabled()
etc..

So how about `app.isPackaged()` and `remote.app.isPackaged()`?

**Comment by ungoldman** [comment]
Hmm, responded via email earlier but it looks like it didn't go through...

I thought about this last night and was about to suggest `app.isPackaged()` as being more in line with the current API. So re: @enlight and @zeke's previous comments: 👍

**Comment by ungoldman** [comment]
I thought about this last night and was about to suggest app.isPackaged()
as being more in line with current API 😂

**Comment by parro-it** [comment]
The `app.isPackaged()` solution has the same disadvantage of the actual `electron-is-dev` implementation: some tools and apps are in production state but they are not packaged.

I would prefer the solution proposed by @MarshallOfSound: seems more flexible and has the advantage that the state can be changed while an app is running.

However, also the `isPackaged` method is useful, normal app could set theirs state doing:

```js
app.setProduction(app.isPackaged());
```

while tools that ships unpackaged could do what @MarshallOfSound said:

```js
app.setProduction(process.env.ELECTRON_ENV === 'production');
```

or whatever they want...

**Comment by lukeed** [comment]
FWIW, I'd much prefer a binary system. It's a little more concise & accepting a string suggests that `development` is just as welcome as `production`. If `dev` is already the default mode, there's no reason to check if `ENV === 'dev'` everywhere; it's assumed until made otherwise.

```json
{
  "scripts": {
    "start": "ELECTRON_PROD=1 electron ."
  }
}
```

Could just as easily be `ELECTRON_DEV`, whichever you think should _not_ be run by default.

```js
app.isProduction = () => process.env.ELECTRON_PROD || false;
```

**Comment by deepak** [comment]
tl;dr: have NODE_ENV and ELECTRON_ENV. and both and almost synonymous 

i like how many libraries have a `DEBUG` environment variable and we can just start the app with 
`DEBUG=* npm start` to see all the messages. it is a standard (ish) and everything is good with the world :-)

something similar for dev/production would be nice. so rather than `NODE_ENV` or `ELECTRON_ENV` 
we just have one flag maybe `APP_ENV=development|production|staging`

do not think this will happen :-) 
unless everyone changes over from `NODE_ENV` to `APP_ENV` or some-such

also i think, it still might be a good idea to set node and electron env seperately. so

1. if there is only `NODE_ENV`, it sets `ELECTRON_ENV`. so `NODE_ENV=production` automatically sets `ELECTRON_ENV=production` as well
2. can override `ELECTRON_ENV`. so `ELECTRON_ENV=development NODE_ENV=production` is valid

also not sure what to do when only `ELECTRON_ENV` is used. 
ie. should `ELECTRON_ENV=production` set `NODE_ENV=production` as well ? (yes, i think) 

we may still need to set `NODE_ENV` even with `ELECTRON_ENV` because electron can and will use nodejs libraries.

not sure where different NODE_ENV and ELECTRON_ENV would be useful 
maybe to give a custom development build

btw [Nylas N1](https://github.com/nylas/N1/blob/0fd751d40363ac1ce58cfc9313bdc18350faeb79/package.json#L109) uses a `--dev` flag to start Electron. But is not documented in the [electron command-line flags](https://github.com/electron/electron/blob/master/docs/api/chrome-command-line-switches.md). Not sure where it is coming from ? cc @emorikawa

**Comment by MarshallOfSound** [comment]
@deepak 

> btw Nylas N1 uses a --dev flag to start Electron.

The `--dev` flag is custom and read in from `process.argv`, I also use `--dev` in GPMDP and read it from `process.argv` to determine if I am in "development mode" or not.  It is not related to the Chromium CLI flags

**Comment by paulcbetts** [comment]
@zeke how about:

```js
let isDevMode = process.execPath.match(/electron[\\/]electron/i)
```

**Comment by ungoldman** [comment]
Old thread, but after above discussion I still think `app.isPackaged()` is the way to go for now. It's a meaningful thing electron can easily return.

`NODE_ENV`/`ELECTRON_ENV`, though useful conventions, are arbitrary environmental variables that should be managed by users.

> re: Where is NODE_ENV documented (nodejs google group, 2012)
>
> It's not documented in node core because it's just a convention, made
popular by express. Node itself doesn't care what the environment vars
are really, it just lets you read and write them via
http://nodejs.org/docs/latest/api/process.html#process_process_env .

- https://groups.google.com/d/msg/nodejs/GGL-SGfWKkI/iz9qBYUysVIJ

DEBUG is also a useful convention, but was something introduced in userland (also by tj of express fame) https://www.npmjs.com/package/debug.

**Comment by myst729** [comment]
I personally find the proposal from @parro-it makes sense. Also agreed with @ungoldman that `ELECTRON_ENV` sounds like something controlled by the end user.

**Comment by zeke** [comment]
Dusting off this thread. 🌬 

Reading back through, there are two basic points we're all mostly agreeing on:

1. Environment variables like `NODE_ENV` and `ELECTRON_ENV` should be set by users, not by Electron.
1. A new `app.isPackaged()` method would be useful.

Implementation-wise, how can one reliably determine whether an app is packaged, regardless of ASAR settings, operating system, etc?

**Comment by zackshapiro** [comment]
Hey @zeke, was there ever a PR here to change behavior? Thanks!

**Comment by zeke** [comment]
@zackshapiro no this never turned into a PR, but it still should! :)

Are you interested in working on it? If not, I can eventually get to it.

**Comment by MarshallOfSound** [comment]
Just because I saw this flying through my inbox, I think the most reliable way to determine if an app is "packaged" or not is to check if `app.asar` or `app` **exists**.  If it does assume prod, if it doesn't assume dev.

This logic is based on the search paths: https://github.com/electron/electron/blob/a6ea316a5de1964761a82d9376c23e99eff16d25/lib/browser/init.js#L106

**Comment by dev-dmtrllv** [comment]
What about `const isDev = process.argv0.includes("node_modules")` ?

**Comment by baconbrad** [comment]
This might be helpful to some. I have previously posted a method we used since 2016. It is no longer a working option in v7.0.X. We have since updated it to pass a dev flag when launching from the `electron` package: (`"electron ./dist/js/main.js --dev"`) 

Our main process checks with the following:

```javascript
function isDev() {
  return process.argv[2] == '--dev';
}
```

Our render process checks with the following:
(Take note `remote` is `require('electron').remote`)
```javascript
function isDev() {
  return remote.process.argv[2] == '--dev';
}
```

Usage:

``` javascript
if (isDev()) {
  // Do dev stuff
}
```

For this we purposely only look for it as the 3rd argv. If your launch script has multiple flags you might have to adjust the specific argv it is looking at. If you want a packaged binary to run this way or make it agnostic to it's placement you best make it loop through `process.argv` looking for a match.

**Comment by ishwarrimal** [comment]
I have noticed that the `process.env` variable in development contains information about `npm_package` which is absent in production build.
Checking for `process.env.npm_node_execpath` gives me info about prod or dev.
Will this be good enough?

**Comment by hayyaun** [comment]
> ## The Problem
> When developing an Electron app, it is sometimes necessary to change the app's behavior depending on whether the app is being run in development or is packaged and running on an end-user's machine. Examples include:
> 
> * Enabling DevTools in "development", disabling in "production"
> * Changing logging/debugging behavior
> * Using different endpoints for talking to webservices
> * Enabling/disabling caching
> 
> ## Proposal
> Electron should provide an intuitive way for users to detect (from any process) whether or not their app is running in a packaged state.
> 
> ## The Current Options
> #### `process.defaultApp`
> Electron has a property called `process.defaultApp` which is set to `true` when running Electron with a target directory, e.g. `electron .`. There are at least two problems with this:
> 
> 1. Only works in the main process. Rendered processes can't access this property by default.
> 2. The name `defaultApp` is not intuitive, and the [docs give no indication](http://electron.atom.io/docs/all/#processdefaultapp) of this property's practical use.
> 
> #### `electron-is-dev`
> [electron-is-dev](https://github.com/sindresorhus/electron-is-dev/) is a userland module that solves this problem. This modules is great, but the issue here is discoverability: Users have to know about it before they can use it.
> 
> #### `ELECTRON_ENV`
> See [electron-userland/electron-prebuilt#118](https://github.com/electron-userland/electron-prebuilt/pull/118), an open PR that aims to set `process.env.ELECTRON_ENV` to `development` in development.
> 
> One might argue that this technique could be used by simply prefixing the environment variable to a start script:
> 
> ```json
> {
>   "scripts": {
>     "start": "ELECTRON_ENV=development electron ."
>   }
> }
> ```
> 
> Unfortunately this is not a cross-platform solution: it won't work on Windows without bringing in an external dependency like [cross-env](https://github.com/kentcdodds/cross-env).
> 
> ## What to do?
> I'm in favor of the `process.env.ELECTRON_ENV` approach, setting it to `development` when run with `electron some/app/dir`, and leaving it `undefined` for packaged apps.
> 
> I like this solution because it resembles the `NODE_ENV` convention from Node.js, and gives users a familiar way to override the value easily, if needed. I would also like to solve this problem in Electron itself, rather than `electron-prebuilt`, so all developers will experience the same behavior, regardless of how they're running Electron.
> 
> Before jumping into implementation, I would love feedback from users. Does this approach sound good?
> 
> cc @parro-it @sindresorhus @ungoldman @kevinsawicki

You can use "npm i -D shx" to use your command cross-platform.

**Comment by hayyaun** [comment]
> > ## The Problem
> > When developing an Electron app, it is sometimes necessary to change the app's behavior depending on whether the app is being run in development or is packaged and running on an end-user's machine. Examples include:
> 
> You can use "npm i -D shx" to use your command cross-platform.

**Comment by Bosch-Eli-Black** [comment]
To anyone running across this thread: It looks like `electron` now supports `app.isPackaged`: https://github.com/electron/electron/blob/main/docs/api/app.md#appispackaged-readonly 🙂

**Comment by sayhicoelho** [comment]
How about adding an `IS_PACKAGED` environment variable to `process.env`?

**Comment by zhiftyDK** [comment]
I found that the best way to solve it is just using this method:
![image](https://user-images.githubusercontent.com/62703235/207019670-2aabe93a-9007-48da-b7ca-ce1bf40ef73a.png)

**Comment by malshoff** [comment]
> I found that the best way to solve it is just using this method:
> 
> ```js
> //Default base of file path
> let pathBase = process.resourcesPath;
> const development = true;
> if(development){
>     pathBase = "./src";
> }
> console.log(pathBase);
> ```

How is `development` never not true?

**Comment by hayyaun** [comment]
I think you may need sth like this?
```javascript
const basePath = process.env.NODE_ENV === 'development' ? './src' : process.resourcesPath;
```
________________________________

**Comment by navidshad** [comment]
I you use Webpack build, there is a `process.env.WEBPACK_SERVE` which shows True when you start the app. When you build the app this variable disappears. so you can consider this:

```js
const isDev = () => {
	return process.env['WEBPACK_SERVE'];
}
```

**Comment by Jimbly** [comment]
The first response by @baconbrad seemed to be exactly what I wanted (but no longer works), an update to that which works in the current version of Electron is:
```js
function isDev() {
  return !app.getAppPath().includes('app.asar');
}
```
In my case I have a packaged development build (not asar'd, allows debugging/etc) and production build (asar'd, connects to production endpoints), so just `app.isPackaged` was insufficient.
